### ISSUE TITLE
Better error message for float arg to np.histogram bins

### ISSUE THREAD

**Issue by matthew-brett** [numpy/numpy#6469]
In numpy 1.9.x and previous - this worked:

```
In [9]: x = np.arange(10)
In [10]: np.histogram(x, bins=2.1)
```

It looks like np.histogram is casting to int, because bins=2.1 and bins=2.9 both give 2 bins.

With numpy 1.10, I get the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-431364707b24> in <module>()
----> 1 np.histogram(x, bins=2.1)

/Users/mb312/.virtualenvs/test/lib/python2.7/site-packages/numpy/lib/function_base.pyc in histogram(a, bins, range, normed, weights, density)
    247                 n.imag += np.bincount(indices, weights=tmp_w.imag, minlength=bins)
    248             else:
--> 249                 n += np.bincount(indices, weights=tmp_w, minlength=bins).astype(ntype)
    250 
    251         # We now compute the bin edges since these are returned

ValueError: operands could not be broadcast together with shapes (2,) (3,) (2,) 
```

I can see the argument for not allowing floats that are not integers, but maybe a more explicit error message would help?

Found via : https://github.com/nipy/nipy/issues/369

**Comment by astrofrog** [comment]
I agree that this should either be fixed to convert to int, or raise a better message. It has personally helped me to find bugs in code so I'd actually go for an error, but it won't be backward-compatible.

**Comment by inesw** [comment]
This has been fixed in 34b582aadae8272e7b7209f7a05594e9258ba217 .
The issue can be closed.

**Comment by charris** [comment]
Thanks for the heads up.
